[SPARK-26666][SQL] Support DSv2 overwrite and dynamic partition overwrite.

rdblue opened the pull request:
## What changes were proposed in this pull request?

This adds two logical plans that implement the ReplaceData operation from the [logical plans SPIP](https://docs.google.com/document/d/1gYm5Ji2Mge3QBdOliFV5gSPTKlX4q1DCBXIkiyMv62A/edit?ts=5a987801#heading=h.m45webtwxf2d). These two plans will be used to implement Spark's `INSERT OVERWRITE` behavior for v2.

Specific changes:
* Add `SupportsTruncate`, `SupportsOverwrite`, and `SupportsDynamicOverwrite` to DSv2 write API
* Add `OverwriteByExpression` and `OverwritePartitionsDynamic` plans (logical and physical)
* Add new plans to DSv2 write validation rule `ResolveOutputRelation`
* Refactor `WriteToDataSourceV2Exec` into trait used by all DSv2 write exec nodes

## How was this patch tested?

* The v2 analysis suite has been updated to validate the new overwrite plans
* The analysis suite for `OverwriteByExpression` checks that the delete expression is resolved using the table's columns
* Existing tests validate that overwrite exec plan works
* Updated existing v2 test because schema is used to validate overwrite

rdblue committed:
DSv2: Support overwrite and dynamic partition overwrite.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101497/
Test FAILed.

rdblue commented:
@mccheah, @felixcheung, and @cloud-fan, can you review this? It adds overwrite plans for DSv2.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7293/
Test FAILed.

SparkQA commented:
**[Test build #101498 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101498/testReport)** for PR 23606 at commit [`5766c1f`](https://github.com/apache/spark/commit/5766c1f52171608c072b89492840c4f72f22328f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7294/
Test PASSed.

SparkQA commented:
**[Test build #101498 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101498/testReport)** for PR 23606 at commit [`5766c1f`](https://github.com/apache/spark/commit/5766c1f52171608c072b89492840c4f72f22328f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class AppendData(`
  * `case class OverwriteByExpression(`
  * `case class OverwritePartitionsDynamic(`
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `
  * `case class AppendDataExec(`
  * `case class OverwriteByExpressionExec(`
  * `case class OverwritePartitionsDynamicExec(`
  * `case class WriteToDataSourceV2Exec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101498/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7295/
Test PASSed.

SparkQA commented:
**[Test build #101499 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101499/testReport)** for PR 23606 at commit [`233d5c5`](https://github.com/apache/spark/commit/233d5c502d13e423f3e125e3c0aaa0131b79c7b5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class AppendData(`
  * `case class OverwriteByExpression(`
  * `case class OverwritePartitionsDynamic(`
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `
  * `case class AppendDataExec(`
  * `case class OverwriteByExpressionExec(`
  * `case class OverwritePartitionsDynamicExec(`
  * `case class WriteToDataSourceV2Exec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101499/
Test FAILed.

mccheah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -386,16 +385,64 @@ case class AppendData(
+
+object OverwriteByExpression {
+  def byName(
+      table: NamedRelation, df: LogicalPlan, deleteExpr: Expression): OverwriteByExpression = {
```
Is there a reason it's `df` here and `query` in `byPosition`?

rdblue replied:
Yes. The DataFrame API uses `byName`, while the SQL path uses `byPosition`.

mccheah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -386,16 +385,64 @@ case class AppendData(
+    isByName: Boolean) extends V2WriteCommand
+
+object OverwritePartitionsDynamic {
+  def byName(table: NamedRelation, df: LogicalPlan): OverwritePartitionsDynamic = {
```
Similarly here. Should the parameter names be consistent?

rdblue replied:
See my other comment. This signals that this corresponds to the behavior of DataFrames.

mccheah reviewed (approved):
I didn't see very many concerns, overall this looks very good. Had a minor comment about a var name.

mccheah commented:
Also needs to pass tests and style of course.

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -365,14 +365,13 @@ case class Join(
-    table: NamedRelation,
-    query: LogicalPlan,
-    isByName: Boolean) extends LogicalPlan {
+private[sql] trait V2WriteCommand extends Command {
```
catalyst is already a private module. so it doesn't need `private[sql]`.

rdblue replied:
I'm not sure that's true. Data types are defined in catalyst, but are part of the public API, right? Is it okay to leave this as-is to be safe?

rdblue replied:
Okay, I can remove it. But I don't think it makes sense considering that this can't actually be a private module. Did I miss something there?

rdblue replied:
Agreed. I'll stick to the convention for now.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/SupportsOverwrite.java:
```diff
@@ -0,0 +1,36 @@
+ * Overwriting data by filter will delete any data that matches the filter and replace it with data
+ * that is committed in the write.
+ */
+public interface SupportsOverwrite extends WriteBuilder {
```
ah i see, make senses. So this `SupportsOverwrite` is mostly for partitioned tables, though it doesn't mention partition at all. I think the semantic can also apply to non-partitioned tables but that will be very hard to implement.

cloud-fan replied:
BTW for a normal INSERT OVERWRITE, which needs to truncate the entire table, the `filters` will be a single true literal?

rdblue replied:
> I think the semantic can also apply to non-partitioned tables but that will be very hard to implement.

Not necessarily. JDBC sources can implement this fairly easily, and those that support transactions can make it an atomic operation. There are also strategies that can work for unpartitioned tables, like deleting data files using min/max ranges show all rows are matched by a filter.

> BTW for a normal INSERT OVERWRITE, which needs to truncate the entire table, the filters will be a single true literal?

I think that truncate is slightly different. Because it is fairly easy to support truncate, but not overwrite by expression, I think that truncate should be a separate operation in the v2 API. I would make `SupportsOverwrite` implement `SupportsTruncate` with a default that calls overwrite with `true` like you suggest, but I think we will need to add a `true` filter.

Also, what do you mean by "normal INSERT OVERWRITE"? What operation is that? Right now, `INSERT OVERWRITE` is effectively dynamic partition overwrite. Unpartitioned tables are truncated because they have just one "partition". Do you agree with that summary?

cloud-fan replied:
> Unpartitioned tables are truncated because they have just one "partition"

If we go with this definition, do you mean data sources that do not support partition must implement `SupportsDynamicOverwrite` to support `INSERT OVERWRITE`?

rdblue replied:
@cloud-fan, I've added `SupportsTruncate` to address this. Now tables can support truncation to implement `INSERT OVERWRITE` that replaces all data in the table.

I also ran tests to find out how file system data sources behave with overwrite mode in `DataFrameWriter`. That showed that tables are always truncated and the value of `spark.sql.sources.partitionOverwriteMode` is always ignored.

rdblue commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7366/
Test PASSed.

SparkQA commented:
**[Test build #101603 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101603/testReport)** for PR 23606 at commit [`233d5c5`](https://github.com/apache/spark/commit/233d5c502d13e423f3e125e3c0aaa0131b79c7b5).

SparkQA commented:
**[Test build #101603 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101603/testReport)** for PR 23606 at commit [`233d5c5`](https://github.com/apache/spark/commit/233d5c502d13e423f3e125e3c0aaa0131b79c7b5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class AppendData(`
  * `case class OverwriteByExpression(`
  * `case class OverwritePartitionsDynamic(`
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `
  * `case class AppendDataExec(`
  * `case class OverwriteByExpressionExec(`
  * `case class OverwritePartitionsDynamicExec(`
  * `case class WriteToDataSourceV2Exec(`

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2245,6 +2245,26 @@ class Analyzer(
         }
+
+      case overwrite @ OverwriteByExpression(table, _, query, isByName)
+        if table.resolved && query.resolved && !overwrite.resolved =>
```
nit: indentation.

rdblue replied:
How should this be indented? 4 spaces from the start of `case`?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2245,6 +2245,26 @@ class Analyzer(
+        }
+
+      case overwrite @ OverwritePartitionsDynamic(table, query, isByName)
+        if table.resolved && query.resolved && !overwrite.resolved =>
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:
```diff
@@ -144,14 +146,22 @@ object DataSourceV2Strategy extends Strategy {
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = splitConjunctivePredicates(deleteExpr).map {
+        filter => DataSourceStrategy.translateFilter(deleteExpr).getOrElse(
+          throw new SparkException(s"Cannot translate expression to source filter: $filter"))
```
Maybe, `AnalysisException` is clearer than `SparkException`?

rdblue replied:
Conversion to physical plan happens after the analysis is finished, so I think the right exception to use is SparkException.

dongjoon-hyun replied:
@rdblue .
I know the logic, but we already use `AnalysisException` in this file. It's based on the characteristic of the exception instead of the location. Please see what you removed.
- https://github.com/apache/spark/pull/23606/files#diff-fc4a83061d40a242e7f7b729cb8ee870L153

rdblue replied:
@cloud-fan, what is the correct exception to use during conversion to a physical plan?

rdblue replied:
`AnalysisException` is not currently referenced in `DataSourceV2Strategy`. I also checked `DataSourceStrategy` and that does not throw `AnalysisException` except in the `DataSourceAnalysis` rules that are run by the analyzer.

An alternative is to consider this a failure to produce a plan and return `Nil`. But that would result in an unhelpful assertion failure, which is worse than a specific `SparkException`.

I think this is doing the right thing.

cloud-fan replied:
Agree that `AnalysisException` is not proper here, as this is not the analysis phase. But `SparkException` is worse. `SparkException` should be used at executor side, which indicates a Spark job is launched.

One thing we can do is, do this conversion at analysis phase. i.e. the logical plan `OverwriteByExpression` should take `Array[Filter]` as well. However, the downside is, for advanced users that touch catalyst rules directly, they are not able to support arbitrary expressions.

So I think `AnalysisException` is OK here, as this is something we should do at the analysis phase, but we don't for other reasons.

rdblue replied:
I'll change this to `AnalysisException`.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,107 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val batchWrite = newWriteBuilder() match {
+      case builder: SupportsSaveMode => // TODO: Remove this
```
It seems that we have three TODOs here. Shall we create SPARK issue and use the ID here explicitly? I guess that all three TODO will have the same SPARK issue ID.

rdblue replied:
Wenchen already opened an issue for this: [SPARK-26356](https://issues.apache.org/jira/browse/SPARK-26356)

dongjoon-hyun replied:
Yes. That's what I requested. We had better have a IDed TODO.
```
// TODO: Remove this
// TODO(SPARK-26356): Remove this
```

rdblue replied:
I'm removing the comments. If SupportsSaveMode is no longer defined, then this has to be removed anyway. The comment doesn't help.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,107 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+        builder.mode(SaveMode.Overwrite).buildForBatch()
+
+      case _ =>
+        throw new SparkException(s"Table does not support dynamic partition overwrite: $table")
```
`AnalysisException`?

rdblue replied:
Other physical plans throw SparkException. I thought AnalysisException was just for query analysis?

dongjoon-hyun replied:
ditto. (https://github.com/apache/spark/pull/23606/files#r251616435)

rdblue replied:
This is definitely a case where it is too late to throw an analysis exception. This will be checked by a new analysis rule when capabilities are added.

cloud-fan replied:
Here we are executing a plan, so `AnalysisException` should not be used. I'm still worried about `SparkException` because we are not launching a Spark job here.

Shall we throw `IllegalStateException`? Eventually we will make this never happen, as this kind of check should be done as early as possible.

rdblue committed:
Update V2 analysis suite to cover overwrite plans.

rdblue committed:
Add SupportsTruncate for sources that do not support overwrite.

SparkQA commented:
**[Test build #101735 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101735/testReport)** for PR 23606 at commit [`013513b`](https://github.com/apache/spark/commit/013513bff586f50f1afa611819f138b07f7666af).

AmplabJenkins commented:
Merged build finished. Test PASSed.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:
```diff
@@ -253,29 +254,38 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
+            case SaveMode.Overwrite =>
+              // truncate the table
+              runCommand(df.sparkSession, "save") {
+                OverwriteByExpression.byName(relation, df.logicalPlan, Literal(true))
```
@gatorsmile, I verified that file system sources truncate in all cases when mode is overwrite. I've updated this to always truncate as you requested.

gatorsmile replied:
-> `filter = Literal(true)`.

cloud-fan replied:
how shall we indicate a truncate? using true literal as the overwrite expression works, another option is to use `Nil`. No overwrite expressions can also mean truncating.

rdblue replied:
@cloud-fan, I think that `Literal(true)` is better than using `Nil`. This translates directly to an `AlwaysTrue` filter, which has obvious behavior: `DELETE FROM table WHERE true`.

Using `Nil` or no filters to signal that all rows should be deleted is not obvious and requires more documentation for the contract with implementations.

cloud-fan replied:
Actually we already have this contract. e.g. `INSERT OVERWRITE tbl SELECT ...` without WHERE will truncate the table and write new data. If the contract is here for end-users, I think it's OK for data source developers as well.

Anyway I think this shouldn't block this PR and we can discuss it later.

rdblue commented:
@cloud-fan, I've added analysis tests for the new logical plans. Existing tests also validate that the new overwrite plan has the same behavior as using WriteToDataSourceV2. Further tests depend on introducing a catalog API because dynamic and static overwrites other than truncate are only available through the SQL API. And we can't translate SQL into v2 plans until we can update the table resolution rule.

rdblue commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -159,6 +159,7 @@ class Analyzer(
     Batch("Resolution", fixedPoint,
       ResolveTableValuedFunctions ::
       ResolveRelations ::
+      ResolveOutputRelation ::
```
I had to move this higher in the batch so that outputs were resolved before the overwrite expression (see analysis tests for `OverwriteByExpression`). Otherwise, the expression would be resolved against the overwrite's child query.

Another approach is to change how the expression's attributes are resolved to use the table instead of the query. But changing expression resolution to use something other than children seemed too intrusive.

rdblue replied:
Thanks for the information about how to solve this. I'll update it.

rdblue committed:
Fix visibility caught by reviews.

SparkQA commented:
**[Test build #101737 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101737/testReport)** for PR 23606 at commit [`6047f65`](https://github.com/apache/spark/commit/6047f6512cb3d8a8a7146494853c60e4ce4ea022).

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #101735 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101735/testReport)** for PR 23606 at commit [`013513b`](https://github.com/apache/spark/commit/013513bff586f50f1afa611819f138b07f7666af).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class AlwaysTrue() extends Filter `
  * `case class AlwaysFalse() extends Filter `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101735/
Test PASSed.

SparkQA commented:
**[Test build #101737 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101737/testReport)** for PR 23606 at commit [`6047f65`](https://github.com/apache/spark/commit/6047f6512cb3d8a8a7146494853c60e4ce4ea022).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait V2WriteCommand extends Command `
  * `trait BatchWriteHelper `
  * `trait V2TableWriteExec extends UnaryExecNode `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101737/
Test PASSed.

rdblue commented:
@cloud-fan, I've added the missing tests to this PR and tests are passing. I think it is ready for a final review. Thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7547/
Test PASSed.

SparkQA commented:
**[Test build #101848 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101848/testReport)** for PR 23606 at commit [`ab9ed01`](https://github.com/apache/spark/commit/ab9ed0199467c34e79e1ffde993d958dc6fc82c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait V2WriteCommand extends Command `
  * `trait BatchWriteHelper `
  * `trait V2TableWriteExec extends UnaryExecNode `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101848/
Test PASSed.

rdblue commented:
@rxin, if you have time, could you review this?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/sources/filters.scala:
```diff
@@ -218,3 +218,23 @@ case class StringEndsWith(attribute: String, value: String) extends Filter {
+/**
+ * A filter that always evaluates to `true`.
+ */
+case class AlwaysTrue() extends Filter {
```
why it's a case class instead of object?

rdblue replied:
There's an object that extends it just below. I thought it was still a good idea to match the pattern of the other filters and make a class as well as an object, since this will be used by Java code. Checking `instanceof AlwaysTrue$` leaks Scala details into Java.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/SupportsOverwrite.java:
```diff
@@ -0,0 +1,42 @@
+ * Overwriting data by filter will delete any data that matches the filter and replace it with data
+ * that is committed in the write.
+ */
+public interface SupportsOverwrite extends WriteBuilder, SupportsTruncate {
```
can we add it later? This PR does not show how it will be used, and only uses `SupportsTruncate`

cloud-fan replied:
nvm, I got it. Both truncate and overwriteByExpression will be handled by one plan: OverwriteByExpression

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+        builder.mode(SaveMode.Overwrite).buildForBatch()
+
+      case _ =>
+        throw new SparkException(s"Table does not support dynamic partition overwrite: $table")
```
we should fail earlier, at analysis time, or at least planning time.

rdblue replied:
Can we do this as a follow-up? I think this check should use capabilities and I don't want to add that API in this commit. Capabilities will be my next PR.

cloud-fan replied:
sure, SGTM

rdblue replied:
Opened [SPARK-26811](https://issues.apache.org/jira/browse/SPARK-26811) to track earlier analysis exceptions.

rdblue committed:
Ignore unresolved writes in ResolveReferences.

This ensures that the query and table columns for a write are resolved
before expressions are resolved so that expressions are resolved using
field names from the table, not the query.

This also reverts moving ResolveOutputRelation because rule order no
longer matters.

rdblue committed:
Remove TODO comments that aren't helpful.

rdblue commented:
@cloud-fan, @dongjoon-hyun, I've updated this PR and replied to the remaining review comments.

I updated `ResolveReferences` as you suggested and the analysis suite is passing tests without relying on rule order. Thank you for your help here, @cloud-fan.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7679/
Test PASSed.

SparkQA commented:
**[Test build #102003 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102003/testReport)** for PR 23606 at commit [`328b57a`](https://github.com/apache/spark/commit/328b57afe3b5abd966fa7a8f83f0e2ee4c9a43dd).

SparkQA commented:
**[Test build #102003 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102003/testReport)** for PR 23606 at commit [`328b57a`](https://github.com/apache/spark/commit/328b57afe3b5abd966fa7a8f83f0e2ee4c9a43dd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102003/
Test FAILed.

rdblue commented:
Retest this please.

Looks like an unrelated, possibly flaky failure: `Caused by: sbt.ForkMain$ForkError: java.lang.ClassNotFoundException: javax.jdo.JDOException`

rdblue commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7697/
Test PASSed.

SparkQA commented:
**[Test build #102026 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102026/testReport)** for PR 23606 at commit [`328b57a`](https://github.com/apache/spark/commit/328b57afe3b5abd966fa7a8f83f0e2ee4c9a43dd).

rdblue commented:
@dongjoon-hyun, there are ORC test failures that look unrelated to me because the v1 tests are also failing with the same problem: cannot infer schema. Do you agree that this is unrelated?

rdblue commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7708/
Test PASSed.

SparkQA commented:
**[Test build #102040 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102040/testReport)** for PR 23606 at commit [`328b57a`](https://github.com/apache/spark/commit/328b57afe3b5abd966fa7a8f83f0e2ee4c9a43dd).

SparkQA commented:
**[Test build #102040 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102040/testReport)** for PR 23606 at commit [`328b57a`](https://github.com/apache/spark/commit/328b57afe3b5abd966fa7a8f83f0e2ee4c9a43dd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102040/
Test FAILed.

rdblue commented:
Retest this please.

rdblue commented:
I think the ORC test failures are unrelated. The same commit passed all but one flaky test in [run #102003](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102003/testReport).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7722/
Test PASSed.

gatorsmile commented:
will do the review in the next few days.

SparkQA commented:
**[Test build #102055 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102055/testReport)** for PR 23606 at commit [`328b57a`](https://github.com/apache/spark/commit/328b57afe3b5abd966fa7a8f83f0e2ee4c9a43dd).

rdblue commented:
The ORC tests are failing due to commit df4c53e44bc9837a470ec66486237403868cb04f by @gengliangwang.

That commit adds a v2 implementation for ORC and uses it by default, which breaks because Overwrite is implemented in this PR using a v2 plan. In v2, overwrite does not create tables, CTAS creates tables. But CTAS cannot be implemented until a table catalog is available.

The solution is to default ORC to the v1 implementation until CTAS is available in v2.

rdblue committed:
SPARK-26666: Use ORC v1 by default.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7723/
Test PASSed.

SparkQA commented:
**[Test build #102056 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102056/testReport)** for PR 23606 at commit [`360e586`](https://github.com/apache/spark/commit/360e5865cb4b0e6b008e84386182a6e2670cccd4).

SparkQA commented:
**[Test build #102055 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102055/testReport)** for PR 23606 at commit [`328b57a`](https://github.com/apache/spark/commit/328b57afe3b5abd966fa7a8f83f0e2ee4c9a43dd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102055/
Test FAILed.

rdblue commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7733/
Test PASSed.

SparkQA commented:
**[Test build #102066 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102066/testReport)** for PR 23606 at commit [`360e586`](https://github.com/apache/spark/commit/360e5865cb4b0e6b008e84386182a6e2670cccd4).

SparkQA commented:
**[Test build #102066 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102066/testReport)** for PR 23606 at commit [`360e586`](https://github.com/apache/spark/commit/360e5865cb4b0e6b008e84386182a6e2670cccd4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102066/
Test FAILed.

rdblue commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7735/
Test PASSed.

SparkQA commented:
**[Test build #102068 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102068/testReport)** for PR 23606 at commit [`360e586`](https://github.com/apache/spark/commit/360e5865cb4b0e6b008e84386182a6e2670cccd4).

SparkQA commented:
**[Test build #102068 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102068/testReport)** for PR 23606 at commit [`360e586`](https://github.com/apache/spark/commit/360e5865cb4b0e6b008e84386182a6e2670cccd4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102068/
Test PASSed.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+        builder.truncate().buildForBatch()
+
+      case builder: SupportsOverwrite =>
+        builder.overwrite(filters).buildForBatch()
```
This requires multiple test cases. 
- an empty filter list
- True
- False
- Null
- And/Or 
- some Filters that the data source does not support? e.g. `StringStartsWith`
- ...

rdblue replied:
I agree that these tests should be added, but there isn't a good way to add them right now. My plan is to add these tests as functionality is exposed in improved APIs that make overwrite by expression available, along with a new test table implementation. Since this PR is already large and there is no way to access this through a public API, I think adding this can wait.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+  }
+}
+
+case class OverwritePartitionsDynamicExec(
```
This needs a class description. Not many readers know what are overwrite partition dynamic

rdblue replied:
Added scaladoc for append, overwrite, and this one.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2237,14 +2242,34 @@ class Analyzer(
         }
+
+      case overwrite @ OverwriteByExpression(table, _, query, isByName)
+        if table.resolved && query.resolved && !overwrite.writeResolved =>
```
Nit: two more space

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2237,14 +2242,34 @@ class Analyzer(
+        }
+
+      case overwrite @ OverwritePartitionsDynamic(table, query, isByName)
+        if table.resolved && query.resolved && !overwrite.writeResolved =>
```
Nit: two more space

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1446,7 +1446,7 @@ object SQLConf {
       " sources will fall back to the V1 sources.")
     .stringConf
-    .createWithDefault("")
+    .createWithDefault("orc")
```
@gengliangwang @cloud-fan Please discuss it with @rdblue and see whether we can keep using data source v2 API even if the table catalog is not ready.

rdblue replied:
This is required. We cannot move sources to v2 by default until we are confident that they will behave the same way. We can't do that until v2 is actually implemented.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -386,16 +387,66 @@ case class AppendData(
+/**
+ * Dynamically overwrite partitions in an existing table.
+ */
+case class OverwritePartitionsDynamic(
```
Users are unable to generate this plan via DF writer APIs?

rdblue replied:
Yes, remember that the DataFrameWriter API is ambiguous, partly because it uses SaveMode and partly because it isn't clear what "save" or "saveAsTable" mean. A better plan is to introduce an API that has obvious behavior, like the one proposed in the logical plans SPIP.

In the short term, this plan will be available from SQL because it implements one of the modes for `INSERT OVERWRITE`.

gatorsmile reviewed (commented):
Finished the first round of the code review. End-to-end test cases for dsv2 are still missing for covering the internal changes. Thanks for your work! @rdblue

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -365,16 +365,17 @@ case class Join(
-  override lazy val resolved: Boolean = {
+  override lazy val resolved: Boolean = writeResolved
+
+  def writeResolved: Boolean = {
```
`inputPlansResolved`?

rdblue replied:
Why?

When you ask for changes, please include your rationale so that I understand the motivation and we can choose an appropriate solution.

cloud-fan replied:
What does `writeResolved` mean? The plans for writing are resolved?

rdblue replied:
It means that the write is resolved: the incoming query matches the output relation. I don't think that `inputPlansResolved` is clear because the output relation is not an input plan.

I'll rename this to `outputResolved`, which I think is more clear.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+
+case class OverwriteByExpressionExec(
+    table: SupportsBatchWrite,
+    filters: Array[Filter],
```
Can you be more clear about what you mean by this? I don't understand what you're asking for.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+        builder.overwrite(filters).buildForBatch()
+
+      case builder: SupportsSaveMode =>
+        builder.mode(SaveMode.Overwrite).buildForBatch()
```
Looks like this cannot support SaveMode. I'll remove it.

rdblue replied:
Because `SaveMode.Overwrite` always truncates, this should be called only if the expression is `AlwaysTrue`. I've updated the code.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
+    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+
+  private def isTruncate(filters: Array[Filter]): Boolean = {
+    filters.length == 1 && filters(0).isInstanceOf[AlwaysTrue]
```
End to end tests should be added later, when there is an implementation that is capable of running them. That would be too large to add to this PR. Like the append PR, this adds tests for the validation rules. I would be happy to build a proper test for v2 plans, but I don't think it is sensible to add it as part of this work.

rdblue replied:
> Return true if one of the filters is AlwaysTrue, right?

No, this uses the same convention as `SupportsPushDownFilters`, where the array of filters should be interpreted as ANDed together. The original expression is split using `splitConjunctivePredicates`.

I've added a note to both `SupportsPushDownFilters` and `SupportsOverwrite` to explicitly note the expected behavior.

> Or we assume the optimizer rule will do it? Do we have an end-to-end test case?

The existing DataFrameWriter overwrite test is an end-to-end test case.

I plan to add more end-to-end test cases when more of this functionality is exposed through SQL and an improved DataFrame write API. Right now, the only case that can be called from the API is table truncation, via `SaveMode.Overwrite`.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:
```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
  */
-case class WriteToDataSourceV2Exec(batchWrite: BatchWrite, query: SparkPlan)
-  extends UnaryExecNode {
+trait V2TableWriteExec extends UnaryExecNode {
```
Can you be more clear about what you are asking?

rdblue replied:
No, v2 exec nodes do not support metrics. I've implemented this in our branch and I can submit a PR to add them, but this PR doesn't change the write logic at all. I think metrics should be added in a follow-up.

rdblue commented on sql/core/src/main/scala/org/apache/spark/sql/sources/filters.scala:
```diff
@@ -218,3 +218,23 @@ case class StringEndsWith(attribute: String, value: String) extends Filter {
+/**
+ * A filter that always evaluates to `false`.
+ */
+case class AlwaysFalse() extends Filter {
```
Will do.

rdblue committed:
SPARK-26666: Update for review comments.

rdblue commented:
@cloud-fan, I've updated this to address comments in the last round of reviews. Please have another look.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7915/
Test PASSed.

SparkQA commented:
**[Test build #102317 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102317/testReport)** for PR 23606 at commit [`bd743d7`](https://github.com/apache/spark/commit/bd743d78461e31c13d3dd6c8279d937b3c39b60c).

SparkQA commented:
**[Test build #102326 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102326/testReport)** for PR 23606 at commit [`fa1fb72`](https://github.com/apache/spark/commit/fa1fb729dfeb2fa56a256f8d250b16b80204acfb).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7922/
Test PASSed.

SparkQA commented:
**[Test build #102317 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102317/testReport)** for PR 23606 at commit [`bd743d7`](https://github.com/apache/spark/commit/bd743d78461e31c13d3dd6c8279d937b3c39b60c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102317/
Test PASSed.

SparkQA commented:
**[Test build #102326 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102326/testReport)** for PR 23606 at commit [`fa1fb72`](https://github.com/apache/spark/commit/fa1fb729dfeb2fa56a256f8d250b16b80204acfb).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102326/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7950/
Test PASSed.

SparkQA commented:
**[Test build #102360 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102360/testReport)** for PR 23606 at commit [`d67ad46`](https://github.com/apache/spark/commit/d67ad46f2090bb0d291c38f31f4bad521a62636c).

SparkQA commented:
**[Test build #102360 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102360/testReport)** for PR 23606 at commit [`d67ad46`](https://github.com/apache/spark/commit/d67ad46f2090bb0d291c38f31f4bad521a62636c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102360/
Test PASSed.

cloud-fan commented:
LGTM except a few minor comments, thanks for working on it!

cloud-fan commented:
BTW it would be good to add end-to-end tests, but since we don't have any implementation yet, I think it's fine to delay it when we fully migrate file source to v2 API.

rdblue committed:
SPARK-26666: Use AnalysisException in DataSourceV2Strategy.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7979/
Test PASSed.

SparkQA commented:
**[Test build #102399 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102399/testReport)** for PR 23606 at commit [`0e42cc2`](https://github.com/apache/spark/commit/0e42cc28b173f55c645c863732fb13d1343d4d75).

SparkQA commented:
**[Test build #102399 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102399/testReport)** for PR 23606 at commit [`0e42cc2`](https://github.com/apache/spark/commit/0e42cc28b173f55c645c863732fb13d1343d4d75).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102399/
Test PASSed.

gengliangwang reviewed (commented):
LGTM

cloud-fan commented:
thanks, merging to master!

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:
```diff
@@ -264,29 +265,38 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
+            case SaveMode.Overwrite =>
+              // truncate the table
+              runCommand(df.sparkSession, "save") {
+                OverwriteByExpression.byName(relation, df.logicalPlan, Literal(true))
```
Sorry if I missed some discussions here but why did we again end up with reading schema here again, for instance via:

```scala
	at org.apache.spark.sql.execution.datasources.orc.OrcUtils$.readSchema(OrcUtils.scala:85)
	at org.apache.spark.sql.execution.datasources.v2.orc.OrcTable.inferSchema(OrcTable.scala:38)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.$anonfun$dataSchema$1(FileTable.scala:45)
	at scala.Option.orElse(Option.scala:306)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.dataSchema$lzycompute(FileTable.scala:45)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.dataSchema(FileTable.scala:44)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.schema(FileTable.scala:53)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$.create(DataSourceV2Relation.scala:100)
```

?

HyukjinKwon replied:
Two tests are being failed by this (https://github.com/apache/spark/commit/4dce45a5992e6a89a26b5a0739b33cfeaf979208 and this were independently merged at similar time range so the tests didn't catch, which is fine - it happens time to time) :

```
org.apache.spark.sql.FileBasedDataSourceSuite.SPARK-24204 error handling for unsupported Interval data types - csv, json, parquet, orc
org.apache.spark.sql.FileBasedDataSourceSuite.SPARK-24204 error handling for unsupported Null data types - csv, parquet, orc
```

```
ScalaTestFailureLocation: org.apache.spark.sql.FileBasedDataSourceSuite at (FileBasedDataSourceSuite.scala:349)
org.scalatest.exceptions.TestFailedException: Expected exception org.apache.spark.sql.AnalysisException to be thrown, but org.apache.spark.SparkException was thrown
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:528)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:527)
	at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1560)
	at org.scalatest.Assertions.intercept(Assertions.scala:812)
	at org.scalatest.Assertions.intercept$(Assertions.scala:802)
	at org.scalatest.FunSuite.intercept(FunSuite.scala:1560)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$69(FileBasedDataSourceSuite.scala:349)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$68(FileBasedDataSourceSuite.scala:348)
	at org.apache.spark.sql.catalyst.plans.SQLHelper.withSQLConf(SQLHelper.scala:47)
	at org.apache.spark.sql.catalyst.plans.SQLHelper.withSQLConf$(SQLHelper.scala:31)
	at org.apache.spark.sql.FileBasedDataSourceSuite.org$apache$spark$sql$test$SQLTestUtilsBase$$super$withSQLConf(FileBasedDataSourceSuite.scala:37)
	at org.apache.spark.sql.test.SQLTestUtilsBase.withSQLConf(SQLTestUtils.scala:230)
	at org.apache.spark.sql.test.SQLTestUtilsBase.withSQLConf$(SQLTestUtils.scala:228)
	at org.apache.spark.sql.FileBasedDataSourceSuite.withSQLConf(FileBasedDataSourceSuite.scala:37)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$67(FileBasedDataSourceSuite.scala:346)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$67$adapted(FileBasedDataSourceSuite.scala:332)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$66(FileBasedDataSourceSuite.scala:332)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$66$adapted(FileBasedDataSourceSuite.scala:330)
	at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1(SQLTestUtils.scala:75)
	at org.apache.spark.sql.test.SQLTestUtils.$anonfun$withTempDir$1$adapted(SQLTestUtils.scala:74)
	at org.apache.spark.SparkFunSuite.withTempDir(SparkFunSuite.scala:116)
	at org.apache.spark.sql.FileBasedDataSourceSuite.org$apache$spark$sql$test$SQLTestUtils$$super$withTempDir(FileBasedDataSourceSuite.scala:37)
	at org.apache.spark.sql.test.SQLTestUtils.withTempDir(SQLTestUtils.scala:74)
	at org.apache.spark.sql.test.SQLTestUtils.withTempDir$(SQLTestUtils.scala:73)
	at org.apache.spark.sql.FileBasedDataSourceSuite.withTempDir(FileBasedDataSourceSuite.scala:37)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$65(FileBasedDataSourceSuite.scala:330)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:104)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
	at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
	at org.apache.spark.sql.FileBasedDataSourceSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(FileBasedDataSourceSuite.scala:37)
	at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:221)
	at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:214)
	at org.apache.spark.sql.FileBasedDataSourceSuite.runTest(FileBasedDataSourceSuite.scala:37)
	at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
	at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
	at org.scalatest.Suite.run(Suite.scala:1147)
	at org.scalatest.Suite.run$(Suite.scala:1129)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
	at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:53)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:53)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1340)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1334)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1334)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:1031)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1500)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:131)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
Caused by: org.apache.spark.SparkException: Could not read footer for file: file:/private/var/folders/71/484zt4z10ks1vydt03bhp6hr0000gp/T/spark-77cc900f-2c11-4961-a704-8f013e7390b6/files/part-00001-7a740eab-e3b0-4354-83ba-2e3c0423aeb2-c000.csv
	at org.apache.spark.sql.execution.datasources.orc.OrcUtils$.readSchema(OrcUtils.scala:75)
	at org.apache.spark.sql.execution.datasources.orc.OrcUtils$.$anonfun$readSchema$2(OrcUtils.scala:85)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:459)
	at scala.collection.TraversableOnce.collectFirst(TraversableOnce.scala:148)
	at scala.collection.TraversableOnce.collectFirst$(TraversableOnce.scala:135)
	at scala.collection.AbstractIterator.collectFirst(Iterator.scala:1429)
	at org.apache.spark.sql.execution.datasources.orc.OrcUtils$.readSchema(OrcUtils.scala:85)
	at org.apache.spark.sql.execution.datasources.v2.orc.OrcTable.inferSchema(OrcTable.scala:38)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.$anonfun$dataSchema$1(FileTable.scala:45)
	at scala.Option.orElse(Option.scala:306)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.dataSchema$lzycompute(FileTable.scala:45)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.dataSchema(FileTable.scala:44)
	at org.apache.spark.sql.execution.datasources.v2.FileTable.schema(FileTable.scala:53)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$.create(DataSourceV2Relation.scala:100)
	at org.apache.spark.sql.DataFrameWriter.relation$lzycompute$1(DataFrameWriter.scala:268)
	at org.apache.spark.sql.DataFrameWriter.relation$1(DataFrameWriter.scala:268)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:278)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:230)
	at org.apache.spark.sql.FileBasedDataSourceSuite.$anonfun$new$70(FileBasedDataSourceSuite.scala:350)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.scalatest.Assertions.intercept(Assertions.scala:805)
	... 75 more
Caused by: org.apache.orc.FileFormatException: Not a valid ORC file file:/private/var/folders/71/484zt4z10ks1vydt03bhp6hr0000gp/T/spark-77cc900f-2c11-4961-a704-8f013e7390b6/files/part-00001-7a740eab-e3b0-4354-83ba-2e3c0423aeb2-c000.csv (maxFileLength= 9223372036854775807)
	at org.apache.orc.impl.ReaderImpl.extractFileTail(ReaderImpl.java:546)
	at org.apache.orc.impl.ReaderImpl.<init>(ReaderImpl.java:370)
	at org.apache.orc.OrcFile.createReader(OrcFile.java:342)
	at org.apache.spark.sql.execution.datasources.orc.OrcUtils$.readSchema(OrcUtils.scala:62)
	... 95 more

```

cloud-fan replied:
that's weird, orc v2 is disabled in this PR.

Let me try these 2 test locally, and revert this PR if needed.

HyukjinKwon replied:
Ah, thanks. I just opened a PR to disable the tests themselves. I will close mine if there's another fix soon.

HyukjinKwon replied:
ah, looks the test explicitly sets `spark.sql.sources.write.useV1SourceList`.
